Add LwM2M firmware push/pull function

rbtchc opened the pull request:
Implement firmware push and pull functions
- Add a state machine for handling the firmware update process (both push & pull)
- Use http_parser to parse URI passed from the server

rbtchc committed:
net: lwm2m: add state machine for firmware pull update

1. Parse firmware pull URI
2. Add lwm2m_firmware_get/set_update_cb() for applicaiton to register
   callback. This is because we want to check the update_state before
   we pass to the application
3. Add lwm2m_firmware_get/set_update_result() and
   lwm2m_firmware_get/set_update_stat() to manage the state trasnition
   as well as the sanity check

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc committed:
net: lwm2m: add firmware push support

1. Add handling block1 option in handle_request(). The basic idea is
   to declare structure block_context at compiled time and use "token"
   as a key to pick up the on-going block cotext. It should be able to
   support multiple blockwise transfer concurrently
2. Use write callback implemented in lwm2m_obj_firmware to deal w/ the
   update state transition and than call the callback registered by the
   application
3. move default_block_size to lwm2m_engine.c to share btw lwm2m_engine
   and lwm2m_obj_firmware_pull

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -92,14 +101,40 @@ static int transfer_request(struct zoap_block_context *ctx,
+	/* if path is not available, off/len will be zero */
+	off = parsed_uri.field_data[UF_PATH].off;
+	len = parsed_uri.field_data[UF_PATH].len;
+	cursor = firmware_uri+off;
```
Please put space before and after the '+'

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -209,32 +244,91 @@ static void firmware_transfer(struct k_work *work)
+
+	off = parsed_uri.field_data[UF_HOST].off;
+	len = parsed_uri.field_data[UF_HOST].len;
+	/* IPv6 is wrapped by brackets */
```
Please put empty line before the comment, looks a bit better that way.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -209,32 +244,91 @@ static void firmware_transfer(struct k_work *work)
+	} else {
+		family = AF_INET;
+		/* Distinguish IPv4 or DNS */
+		for (int i = off; i < off+len; i++) {
```
spaces missing before and after +

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -209,32 +244,91 @@ static void firmware_transfer(struct k_work *work)
+	}
+
+	tmp = firmware_uri[off+len];
+	firmware_uri[off+len] = '\0';
```
same issue here with '+', can you do a search and fix all of these in one go

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1902,20 +1909,92 @@ static int do_write_op(struct lwm2m_engine_obj *obj,
+	if (ctx == NULL) {
+		return;
+	}
+	ctx->tkl = 0;
```
empty line after }

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1933,6 +2012,7 @@ static int handle_request(struct zoap_packet *request,
 	struct lwm2m_engine_context context;
 	int observe = -1; /* default to -1, 0 = ENABLE, 1 = DISABLE */
 	bool discover = false;
+	struct block_context *ctx = NULL;
```
There is already context variable, and now you introduced ctx. These look very similar to my eyes, could you rename this one to something like block_ctx or similar in order to avoid confusion.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2046,7 +2125,19 @@ static int handle_request(struct zoap_packet *request,
+		if (r < 0) {
+			goto cleanup;
+		}
+		context.block_offset = zoap_next_block(in.in_zpkt, &ctx->ctx);
```
empty line after }

jukkar commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2502,6 +2615,9 @@ int lwm2m_engine_start(struct net_context *net_ctx)
 static int lwm2m_engine_init(struct device *dev)
 {
+	memset(block1_contexts, 0,
+	       sizeof(struct block_context)*NUM_BLOCK1_CONTEXT);
```
add spaces before and after '*'

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,22 +60,117 @@ static struct lwm2m_engine_obj_inst inst;
+		assert(0 && "Shouldn't happen");
+		break;
+	}
+	update_result = result;
```
empty line after }

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -99,6 +204,45 @@ lwm2m_engine_set_data_cb_t lwm2m_firmware_get_write_cb(void)
+		SYS_LOG_ERR("State other than downloaded: %d", state);
+		return -1;
+	}
+	lwm2m_firmware_set_update_state(STATE_UPDATING);
```
empty line after }

jukkar commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -653,14 +653,17 @@ int do_write_op_tlv(struct lwm2m_engine_obj *obj,
-	while (tlvpos < in->insize) {
-		len = oma_tlv_get(&tlv, &in->inbuf[tlvpos],
-				   in->insize - tlvpos);
+	u8_t *inbuf = in->inbuf;
```
I looks like you have extra empty line between variable declarations above.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -83,7 +158,16 @@ static int package_uri_write_cb(u16_t obj_inst_id,
+		lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+		lwm2m_firmware_start_transfer(package_uri);
+	} else if (STATE_DOWNLOADED == state) {
+		if (data_len == 0) {
```
From [LwM2M specification](https://goo.gl/5HYazk) figure 29, it looks like it only reset to state idle when it's currently at state downloaded and an empty package URI/resource is written.
That's why I put the state == STATE_DOWNLOADED check here.

If we reset the state machine at any state, then we have to do (1) cancel the firmware download and (2) cancel the firmware update.

I'm not sure which is the correct behavior. But from my point of view, it's simpler if we only accept the state machine reset at DOWNLOADED state.

rsalveti replied:
Figure 29 seems to be a possible implementation indeed, but that doesn't cover the cancel update case, which is why my interpretation was more to reset at any point in time, at any state.

Looks like something to ask oma for clarification.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -83,7 +158,16 @@ static int package_uri_write_cb(u16_t obj_inst_id,
+			lwm2m_firmware_set_update_state(STATE_IDLE);
+			/* TODO: clean-up downloaded package? */
+		}
+	}
```
I'm afraid that I did not catch your point.
What kind of checks do I have to add to package_uri_write_cb?

rsalveti replied:
At least to reset the state machine if there is an empty write.

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -99,6 +183,46 @@ lwm2m_engine_set_data_cb_t lwm2m_firmware_get_write_cb(void)
+		if (ret < 0) {
+			SYS_LOG_ERR("Failed to update firmware: %d", ret);
+			lwm2m_firmware_set_update_result(
+					RESULT_INTEGRITY_FAILED);
```
Good idea, I think we can map -EINVAL to RESULT_INTEGRITY_FAILED and other values to RESULT_UPDATE_FAILED. 
There are only three results (success, integrity error, fw update fail) are acceptable according to [LwM2M specification](https://goo.gl/5HYazk) figure 29.

rbtchc replied:
Good suggestion.
I will map -EINVAL to RESULT_INTEGRITY_FAILED and others to RESULT_UPDATE_FAILED.
From [LwM2M specification](https://goo.gl/5HYazk) figure 29, it seems there are only three acceptable results which are RESULT_SUCCESS, RESULT_INTEGRITY_FAILED and RESULT_UPDATE_FAILED.

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2103,16 +2207,37 @@ static int handle_request(struct zoap_packet *request,
 	}
 
+cleanup:
 	if (r == 0) {
```
Since you removed the useless else {}, you could do:

if (r) {
   goto error;
}

and thus all the code block below  could be indented back. Nicer to read, less imbrication of if/else etc...

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2103,16 +2207,37 @@ static int handle_request(struct zoap_packet *request,
+	}
+
+	if (r != 0) {
 		if (r == -ENOENT) {
```
here would be you error label

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -143,13 +143,35 @@ static int package_write_cb(u16_t obj_inst_id,
 {
-	SYS_LOG_DBG("PACKAGE WRITE");
+	u8_t state;
+	int ret = 0;
```
usually, in net stack, we try to have declaration with assignment as first lines, then the rest. So I would reverse the order here.

rbtchc replied:
Delete the debug message directly

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -652,15 +652,17 @@ int do_write_op_tlv(struct lwm2m_engine_obj *obj,
 	struct oma_tlv tlv;
 	int tlvpos = 0, ret;
+	u8_t *inbuf = in->inbuf;
+	u16_t insize = in->insize;
```
declare inbuf and insize before len

rbtchc committed:
net: http: add config to separate URL only parser

Add HTTP_PARSER_URL_ONLY to separate URL only parser

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,11 +60,85 @@ static struct lwm2m_engine_obj_inst inst;
+#if defined(CONFIG_ASSERT)
+	switch (state) {
+	case STATE_DOWNLOADING:
+		assert(update_state == STATE_IDLE);
```
Are we sure that we want to assert here (even if CONFIG_ASSERT is enabled)?  That seems a bit harsh for passing in an invalid state.  Maybe ignore and print an error message?  Especially since this is exposed in application code.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,11 +60,85 @@ static struct lwm2m_engine_obj_inst inst;
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_SUCCESS:
+		assert(update_state == STATE_UPDATING);
```
Once again, I'm not sure an assert is the right behavior here.  Consider a strongly worded error message?

rbtchc replied:
I will replace assert w/ error messages.

rbtchc commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -122,6 +122,7 @@ config LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 	bool "Firmware Update object pull support"
 	default y
 	depends on LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT
+	select HTTP_PARSER
```
I will add a patch to separate the URL parsing from header parsing.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,11 +60,85 @@ static struct lwm2m_engine_obj_inst inst;
+	return update_result;
+}
+
+void lwm2m_firmware_set_update_result(u8_t result)
```
I feel like lwm2m_firmware_set_update_state() and lwm2m_firmware_set_update_result() are performing data_validation and then a followup call to set related data.  In discussions with @rsalveti we had plans to add a new data_validation_callback for data elements which would be called prior to assigning a value to the data element.

This would clean up several issues that I'm seeing here:
- You're creating specific data element based functions to control how a data element is set.  This doesn't scale well over time.
- These functions in turn trigger related data element values to be set.
- In application code we're using these somewhat private functions (which may or may not be available depending on CONFIGs).

Instead, if we had a data_validation callback which could be assigned to a data element.  The validation could be performed there (returning non-zero value when an invalid value is found).  And if a valid value was found the follow up assignments of update_state could be handled in a post_write callback of update_result, etc.

rbtchc replied:
Yup, it looks like a good idea if we could have a data validation callback when application want to update the value of the object resources. (I'm assuming that you are talking about adding callback to  lwm2m_engine_set_xxx() functions.)
Do I create a ticker on JIRA for tracking it or ?

rbtchc commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1890,20 +1906,93 @@ static int do_write_op(struct lwm2m_engine_obj *obj,
+	zoap_block_transfer_init(&(*ctx)->ctx, ZOAP_BLOCK_64, 0);
+#else
+	zoap_block_transfer_init(&(*ctx)->ctx, ZOAP_BLOCK_256, 0);
+#endif
```
A JIRA ticket is created for this https://jira.zephyrproject.org/browse/ZEP-2416.
I suppose we will fix this later on.

rbtchc replied:
It seems that patch 61961cf7374dacb7a8a9419f04ecdfc7707ad5d4 submitted by @rsalveti has alter this to a Kconfig. I will update accordingly

tbursztyka commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -84,6 +88,11 @@ static int transfer_request(struct zoap_block_context *ctx,
 	struct zoap_pending *pending = NULL;
 	struct zoap_reply *reply = NULL;
 	int ret;
+	int i;
```
It's actually ok to do both. No need to enforce grouping

rsalveti commented:
@rbtchc while we get all the firmware update review done, would you mind breaking up the fixes you have in a separated pull request? At least "net: lwm2m: fix registration content format and use plain/text", "net: lwm2m: save accept format in observe_node_data" and "net: lwm2m: fix erroneous TLV write", as they are good fixes that could be merged soon.

rbtchc commented:
Here are the summary of the changes I made.
1. Create separate PRs (#1089, #1090, #1091) for patches independent to firmware push/pull
1. Correct the coding style
1. Add b84b6f45a3a92c0aa6ef34b039792bff9ebe5356 for a smaller URL only parser
1. Add NOTIFY_OBSERVER() call when update_result/update_state is updated. Replace assert w/ error message

TODO
1. Clarify whether firmware update can be cancelled at any state or DOWNLOADED state only
1. File a ticket JIRA for implement data_validation callback(?)

mike-scott commented:
JIRA for data validation callback is here: https://jira.zephyrproject.org/browse/ZEP-2500

rbtchc commented:
Raised firmware update cancellation question on https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/219

Do we need to wait for the clarification before we merge the patches?
Or perhaps I could file a JIRA ticket to track and patch the behavior when clarification is made.

rbtchc commented:
@jukkar @tbursztyka @mike-scott @rsalveti 
Please take a look at the updated patches and let me know if there are any unaddress issues.
Thanks.

rbtchc commented:
Rebase and resolve code conflicts according to the change dcb80f7ab888464d52d3ce5701b521c405fa67ed
Diff: https://hastebin.com/egiliqigiv.hs

pfalcon commented:
> Rebase and resolve code conflicts according to the change dcb80f7

Yep, sorry guys, I meant to ping you about it, seeing that lwm2m code uses, but that slipped. @mike-scott , @rsalveti, @mbolivar  FYI

mike-scott commented:
@rbtchc Testing this PR again today / tomorrow and I'll respond back. But @jukkar may want to discuss the Kconfig option to enable HTTP parser only.

jukkar commented on include/net/http_parser.h:
```diff
@@ -227,6 +230,7 @@ struct http_parser_settings {
 	http_cb      on_chunk_header;
 	http_cb      on_chunk_complete;
 };
+#endif
```
Could add the config option in comment here as there are almost 200 lines inside this define block.
So something like this:
#endif /* !CONFIG_HTTP_PARSER_URL_ONLY */

jukkar commented on subsys/net/lib/http/Kconfig:
```diff
@@ -65,6 +65,14 @@ config HTTP_PARSER
+	default n
+	select HTTP_PARSER
+	help
+	This option enables URL parser part of the http_parser
```
Perhaps just say here "This option only enables URL parser of the http_parser library."

jukkar commented on subsys/net/lib/http/http_parser.c:
```diff
@@ -311,6 +314,7 @@ enum header_states {
 	h_connection_close,
 	h_connection_upgrade
 };
+#endif
```
Add the /* !CONFIG_HTTP_PARSER_URL_ONLY */ here

jukkar commented on subsys/net/lib/http/http_parser.c:
```diff
@@ -382,6 +387,7 @@ int cb_data(struct http_parser *parser, http_data_cb cb, int cb_error,
 
 	return 0;
 }
+#endif
```
Same here

jukkar commented on subsys/net/lib/http/http_parser.c:
```diff
@@ -493,6 +500,7 @@ static struct {
 };
 
 int http_message_needs_eof(const struct http_parser *parser);
+#endif
```
And here

jukkar commented on subsys/net/lib/http/http_parser.c:
```diff
@@ -2624,6 +2633,7 @@ const char *http_errno_description(enum http_errno err)
 
 	return http_strerror_tab[err].description;
 }
+#endif
```
And here

jukkar commented on subsys/net/lib/http/http_parser.c:
```diff
@@ -2933,3 +2944,4 @@ unsigned long http_parser_version(void)
 	       HTTP_PARSER_VERSION_MINOR * 0x00100 |
 	       HTTP_PARSER_VERSION_PATCH * 0x00001;
 }
+#endif
```
And here.

jukkar commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -122,6 +122,7 @@ config LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 	bool "Firmware Update object pull support"
 	default y
 	depends on LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT
+	select HTTP_PARSER_URL_ONLY
```
You need to add "select HTTP_PARSER" here too as when doing "select" in Kconfig, it will not recursively select other options.

jukkar replied:
What about my comment here, is the HTTP_PARSER option selected by some other way? What I mean by this is if you select HTTP_PARSER_URL_ONLY here, it will not select HTTP_PARSER automatically (at least that is how Kconfig worked before) i.e., recursive "select" is not implemented. But it might be that the HTTP_PARSER is selected automatically by some other mean in which case my concern is moot.

rbtchc replied:
I did add "select HTTP_PARSER" at [subsys/net/lib/http/Kconfig#L68](https://github.com/rbtchc/zephyr/blob/ee49c3dbdfff43f3ecd0183cdeeffcc1291a0d1c/subsys/net/lib/http/Kconfig#L68) when config HTTP_PARSER_URL_ONLY is selected.
In this way, when LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT is selected, both HTTP_PARSER & HTTP_PARSER_URL_ONLY will be selected as well.

jukkar replied:
Ok, I missed that.

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -228,78 +300,143 @@ static void firmware_transfer(struct k_work *work)
-	if (firmware_addr.sa_family == AF_INET6) {
+	if (IS_ENABLED(CONFIG_NET_IPV6) && firmware_addr.sa_family == AF_INET6) {
 		ret = net_context_bind(firmware_net_ctx,
 				       (struct sockaddr *)&any_addr6,
```
Isn't there now a compile error if you disable IPv6 as the any_addr6 is only found if IPv6 is enabled?

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -228,78 +300,143 @@ static void firmware_transfer(struct k_work *work)
 
-#if defined(CONFIG_NET_IPV4)
-	if (firmware_addr.sa_family == AF_INET) {
+	if (IS_ENABLED(CONFIG_NET_IPV4) && firmware_addr.sa_family == AF_INET) {
```
And same thing for IPv4

jukkar reviewed (commented):
I am ok with the HTTP URL parser only option, I was actually needing something like this myself.

rbtchc commented:
Update according to @jukkar 's comment. Here's the diff: https://hastebin.com/vovevumeca.cs

rbtchc commented:
Update to take care of error handling when failing to add block1 option to the out packet
Diff: https://hastebin.com/upupiguqub.cs

jukkar commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -228,44 +300,112 @@ static void firmware_transfer(struct k_work *work)
+	len = parsed_uri.field_data[UF_HOST].len;
+
+	/* IPv6 is wrapped by brackets */
+	if (off > 0 && firmware_uri[off - 1] == '[') {
```
BTW, I just submitted #1198 which provides net_ipaddr_parse() function that converts a IPv{4|6} address string to struct sockaddr. That function could probably be used here in order to save some memory, this is just something that could be considered here.

rbtchc replied:
Good to know. I will apply those changes to the PR when they are merged.

jukkar reviewed (dismissed):
LGTM

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1934,20 +1950,113 @@ static int do_write_op(struct lwm2m_engine_obj *obj,
+			break;
+		}
+
+		if (timestamp-block1_contexts[i].timestamp >
```
Spaces around "-":
if (timestamp - block1_contexts[i].timestamp >

mike-scott commented on include/net/lwm2m.h:
```diff
@@ -89,9 +89,15 @@ int lwm2m_device_add_err(u8_t error_code);
+#endif
 
+#if defined(CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT)
+void lwm2m_firmware_set_update_cb(lwm2m_engine_exec_cb_t cb);
```
These 2 functions are in lwm2m_obj_firmware.c and should be moved into the above check for CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_object.h:
```diff
@@ -278,6 +278,7 @@ struct lwm2m_engine_context {
 	struct lwm2m_input_context *in;
 	struct lwm2m_output_context *out;
 	struct lwm2m_obj_path *path;
+	size_t block_offset;
```
I'm curious why have the block_offset here in the engine_context as opposed to placing it in the block_context structure where it can be directly related to the struct zoap_block_context?  Feels like this should be moved.

rbtchc replied:
I saved the block_offset at the lwm2m_engine_context because it will be passed to the "post_write_cb" at lwm2m_engine_write() function.
I could use get_block_ctx() each time when lwm2m_engine_write() is called but I chose to store the block_offset at the context to save the query time.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.h:
```diff
@@ -75,4 +75,11 @@ void lwm2m_udp_receive(struct net_context *ctx, struct net_pkt *pkt,
 
+enum zoap_block_size lwm2m_default_block_size(void);
+
+u8_t lwm2m_firmware_get_update_state(void);
```
These functions would only be available if CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT is enabled.

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2148,37 +2273,60 @@ static int handle_request(struct zoap_packet *request,
 
-	if (r == 0) {
-		/* TODO: Handle blockwise 1 */
+	if (r) {
```
I like the restructuring here where we move the handling of the zoap_header_set_code() calls.  However, in the future, you could break this out as a small patch prior to "net: lwm2m: add firmware push support".  The code movement patches are really easy to review, but complicate reviewing larger patches.

rbtchc replied:
Got it.

rsalveti commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1934,20 +1950,113 @@ static int do_write_op(struct lwm2m_engine_obj *obj,
 
+enum zoap_block_size lwm2m_default_block_size(void)
+{
+	switch (CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_BLOCK_SIZE) {
```
You moved this block from a file that was only built when the option above was defined, so either protect this block or make this config generic.

rbtchc replied:
You are right. I did not notice that.
I will make it a generic config for block transfer shouldn't be limited to firmware update only.

rbtchc commented:
Revised according to @mike-scott and @rsalveti 's review.
Here's the diff: https://hastebin.com/nizubewebe.cs

1. Remove block_offset from lwm2m_engine_context
1. Rename LWM2M_FIRMWARE_UPDATE_PULL_COAP_BLOCK_SIZE to LWM2M_COAP_BLOCK_SIZE. Make it generic to LwM2M
1. Move lwm2m_firmware_set_update_cb() & lwm2m_firmware_get_update_cb() into the #if defined(CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT) code block
1. Add LWM2M_NUM_BLOCK1_CONTEXT to Kconfig

mike-scott commented:
Testing this PR again today.

mike-scott commented:
@rbtchc Was there a specific way you were testing this?  For me on QEMU and using Copper firefox plugin to send a PUT w/ File payload (256 block size) to 5/0/0. I see the following log: https://pastebin.com/raw/PFifFe3h  It's looping at the first write w/ a 0 length.

mike-scott commented:
NOTE: I can successfully test PULL method using Californium server running on port 5685 locally to host a binary.  I send a WRITE to 5/0/1 via Leshan (running on port 5683) "coap://192.0.2.2:5685/large-create/1"

rbtchc committed:
net: lwm2m: setup data_ptr/len for OPAQUE resource when none given

OPAQUE resource type might/might not have data_ptr/data_len setup
depending on the implementation. This introduces an issue that when
an OPAQUE resource is written from the server side, the one w/ none
setup will not be able to pass data to post_write_cb()

Modify to setup data_ptr/data_len as incoming buffer and buffer size

Signed-off-by: Robert Chou <robert.ch.chou@acer.com>

rbtchc commented:
@mike-scott I tested it against Leshan server (branch: [fwupdate](https://github.com/eclipse/leshan/tree/fwupdate)) 
But I can reproduce the error, too. Sorry for not being able to catch the issue.

This is because of the change that I made on last update to remove "block_offset" from the lwm2m_engine_context.

It turns out that zoap_next_block() returns
- block_offset when it's not the last block and
- 0 when it's last block **w/o updating the "current" variable** of "struct zoap_block_context". 
It doesn't set the "current" to zero due to it will need it later on for estimating the block number when adding the block option to the outgoing packet to the caller.

The zero data length on the first write is because the OPAQUE resource doesn't have "data_ptr" & "data_len" setup. This will result in not being able to pass the incoming data/data length to the "post_write_cb" of the given resource. To solve this, I have included a new patch 51a3080919decaa6fac354a2729aef6a39cee24a in the PR to fix the issue.

Here is the [diff link]( https://gist.github.com/rbtchc/ea6580b5226ad8049c1b5f7ef9c68f84) and changes are summarized below
1. Add "block_offset" at handle_request() to determine the last block
1. Response 4.13 to caller payload is larger than the network buffer size
1. Fix not passing data_ptr/data_len to "post_write_cb" 51a3080919decaa6fac354a2729aef6a39cee24a
1. Add passing total_size to "post_write_cb"
1. Relocate get_option_int()

mike-scott commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c:
```diff
@@ -156,67 +201,72 @@ do_firmware_transfer_reply_cb(const struct zoap_packet *response,
 
-	/* TODO: Process incoming data */
+	/* Reach last block if transfer_offset equals to 0 */
+	transfer_offset = zoap_next_block(check_response, &firmware_block_ctx);
```
I tested this again this morning and there seems to be an extra zoap_next_block() call here.  At the end of the transfer this causes an EINVAL error to be generated.

NOTE: The zoap_next_block() call below if payload_len > 0 seems to be the correct one.  If I comment out this first call then everything works correctly (it seems).

rbtchc replied:
You are correct.
But we should remove the second one (the one inside the if (payload_len > 0))
Will update accordingly.

mike-scott commented:
As a side note, I see the WIP PR for CoAP API re-write to support multi-fragment is getting worked on here: https://github.com/zephyrproject-rtos/zephyr/pull/1232

@jukkar @tbursztyka I'm not sure what your recommendation is for this PR.. Should we press on and deal w/ the API changes in 1.10?  Do you think that API will present serious conflicts for how we're doing things here?  I think there's value to getting this PR into 1.9 as it would present a full firmware update method over LwM2M, but I understand if it's already past the deadline and will only cause headaches later.

jukkar commented:
> @jukkar @tbursztyka I'm not sure what your recommendation is for this PR.. Should we press on and deal w/ the API changes in 1.10? Do you think that API will present serious conflicts for how we're doing things here? I think there's value to getting this PR into 1.9 as it would present a full firmware update method over LwM2M, but I understand if it's already past the deadline and will only cause headaches later.

Even thou it is tempting to try to push this to 1.9, I think it is better not to rush with this one. So I suggest that we get this to 1.10 instead.

BTW, @rveerama1 sent a PR #1232 which will support multi-fragment CoAP message. That PR is probably useful for lwm2m support. Please review that PR if possible.

rbtchc commented:
Revised the patch according to @mike-scott 's review
Here's the change: https://gist.github.com/rbtchc/266ea65fded6fcdfa96878ae1fb55311

mike-scott commented:
Re-tested tonight w/ your latest change and it seems that everything works now.  LGTM.

mike-scott commented:
@rbtchc I hate to ask, but do you think this PR could be rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/1255 ?

rbtchc commented:
@mike-scott sure

rsalveti commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,33 +59,185 @@ static struct lwm2m_engine_obj_inst inst;
+		 */
+		lwm2m_firmware_set_update_state(STATE_DOWNLOADING);
+	} else if (state != STATE_DOWNLOADING) {
+		if (data_len == 0 && state == STATE_DOWNLOADED) {
```
I wonder if we should always check for data_len == 0 even when state is not downloading. That is just to protect the code flow as the state can be updated via other parts of the code (even when it shouldn't).

rbtchc replied:
Just to confirm. Are you suggesting that we always check if "data_len == 0" and reset the state back to default?

rsalveti replied:
Yes, my concern is what would happen when data_len == 0 but state is not what we expect to  be (simply because state can be changed independently).

rsalveti commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,33 +59,185 @@ static struct lwm2m_engine_obj_inst inst;
+		if (state == STATE_IDLE) {
+			lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+			lwm2m_firmware_start_transfer(package_uri);
+		} else if (state == STATE_DOWNLOADED && data_len == 0) {
```
Same here (data_len == 0).

rsalveti commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -59,33 +59,185 @@ static struct lwm2m_engine_obj_inst inst;
+			/* reset to state idle and result default */
+			lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+		}
+
```
Mind also adding a error message here in case this gets called when state == DOWNLOADING.

rbtchc replied:
Will print out a error message here.

rbtchc commented:
Update to address @rsalveti 's concerns.  Diff: https://hastebin.com/exixuzaxex.cs
1. Reset update state / result to default when empty string is received by either /5/0/0 or /5/0/1
1. Remove TODO comment
1. Response -EPERM (mapped to 4.05) method not allowed when executing 5/0/2 (update) but state is not downloaded

mike-scott commented:
Thanks for the update @rbtchc I'll try and fold these changes into https://github.com/zephyrproject-rtos/zephyr/pull/1255 (using this as a holding PR for 1.10 changes for now)

mike-scott commented:
Probably safe to close this PR.  I have rebased versions of these patches included in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/4208

rbtchc commented:
Close as @mike-scott mentioned
